eng/build.ps1

```diff
@@ -27,11 +27,19 @@ Param(
   [Parameter(ValueFromRemainingArguments=$true)][String[]]$properties
 )
 
+function Get-Default-Arch() {
+  if ($env:DOTNET_RUNTIME_DEFAULT_ARCH) {
+    return $env:DOTNET_RUNTIME_DEFAULT_ARCH
+  }
+
+  return [System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture.ToString().ToLowerInvariant()
+}
+
 function Get-Help() {
   Write-Host "Common settings:"
   Write-Host "  -arch (-a)                     Target platform: x86, x64, arm, arm64, or wasm."
   Write-Host "                                 Pass a comma-separated list to build for multiple architectures."
-  Write-Host ("                                 [Default: {0} (Depends on your console's architecture.)]" -f [System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture.ToString().ToLowerInvariant())
+  Write-Host ("                                 [Default: {0} (Depends on your console's architecture or the DOTNET_RUNTIME_DEFAULT_ARCH environment variable.)]" -f @(Get-Default-Arch))
```

Comment:
This covers for the top level `build.sh1`, but it does not cover other entrypoints into the dev workflow. For example, it does not cover running `dotnet build ....` directly that is part of number of documented dev workflows. If we want to go with the environment variables as the solution for overriding defaults, should it be respected for all entrypoints into the dev workflow?

Reply:
The peculiarity of `build.sh1` is that it resolves the default architecture from the process architecture of the calling shell. The architecture of `dotnet build ...` is controlled by what SDKs you have installed and the `PATH` environment variable. I've never had issues running the correct sdk from a 32-bit shell.

Reply:
> The peculiarity of build.sh1 is that it resolves the default architecture from the process architecture of the calling shell. 

`build.ps1` resolves the default architecture from the process architecture of the calling powershell.exe. It sounds like that you have the x64 powershell.exe on the PATH in your environment, but you actually want to have the arm64 powershell.exe. Is there arm64 powershell on Windows Arm64?

Reply:
Windows Arm does have native builds for Powershell, however I'm using mingw which only has x64 builds at the moment. A [similar issue](https://github.com/dotnet/runtime/issues/104033#issuecomment-2191779696) with SSH connections to Windows arm devices.

Reply:
Would it be better to workaround the problem by changing the path in your mingw environment to point to native powershell? It would cover this and all other powershell.exe invocations.

Reply:
The mingw environment resolves to the same executable (`%SystemRoot%\system32\WindowsPowerShell\v1.0\powershell.exe`) however when launching that from an x64 parent it appears to be adopting an x64 compatibility mode:

![image](https://github.com/dotnet/runtime/assets/2813363/35cef908-e4c0-499c-b7dc-f1b1af4bddfa)

I wasn't able to find a method for launching powershell.exe in arm64 mode from an x64 process. Modern `pwsh` always runs in the native architecture, however our build script is hardcoding powershell.exe. I suppose I could try to create a `powershell` script that invokes pwsh and add it to my path, but it's a lot of ceremony for an issue impacting multiple users.

Reply:
I think the best way to deal with this problem is by having `b.cmd` script that launches `build.cmd` using `%windir%\sysnative\cmd.exe`. You would be running all native from the get-go.

Reply:
That would probably entail maintaining a separate script for each git worktree that is being created, right?

Reply:
This script can be global. It can relaunch the build.cmd from the current directory if there is one. It can do all sorts of other things based on your personal conventions, like run `dotnet build` if there is .csproj in the current directory.

Reply:
`cmd.exe` seems to be governed by the same behavior as `powershell.exe`. It launches as an x64 compatible process if run from an x64 shell.  The only workaround I could find was running `pwsh -c cmd`.

Another possible solution is to default to the machine architecture rather than the process architecture. I'm not sure if dev workflows or CI depend on the current behavior, or what the historic reasons for adopting this default might be.

Reply:
> This script can be global.

That's all true, my only concern is forcing devs using x64 tools (mingw,ssh) to maintain pet scripts for what is basic behavior (building the product using the native arch).

Reply:
cc @vcsjones who mentioned that is impacted by this as well.

Reply:
> It launches as an x64 compatible process if run from an x64 shell. 

Even if you run it as `%windir%\sysnative\cmd.exe` ? I would expect this to launch native cmd process - it is how it works in the x86 WoW at least.

> using x64 tools (mingw,ssh) to maintain pet scripts for what is basic behavior (building the product using the native arch)

The non-native shells keep you on non-native architecture. It has been always the case. 32-bit x86 cmd shell window on Windows will steer everything you do in that shell towards 32-bit x86. Similarly, x64 shell on Apple Silicon will tend to keep you on x64. Even we put in this workaround in place for build.cmd script, I do not expect doing native development from non-native shell to be a great experience.

Reply:
> Even if you run it as %windir%\sysnative\cmd.exe ? 

That folder doesn't exist on ARM builds. Similar to powershell.exe it's the same executable that adapts depending on the parent architecture:

![image](https://github.com/dotnet/runtime/assets/2813363/b2f5549f-513d-4a2d-b70c-fe63fe5c6a3a)

> The non-native shells keep you on non-native architecture. It has been always the case. 

I understand that this is the case, however like I said this does create problems in the workflows for those that do rely on tools that aren't arm64 native yet. While creating a wrapper script _does_ technically resolve the issue, the dev loop workflow largely relies on muscle memory: I work on x64 and arm64 machines interchangeably so even with the workaround I have to explicitly remember that the current machine is arm64 to make sure the right script/arch parameter is being used. If I don't, I will have lost > 5 minutes waiting on a build that was targeting the wrong architecture. This has been happening frequently enough to me for it to be a problem.

Reply:
> That folder doesn't exist on ARM builds.

That's surprising. It has been a virtual folder that only exists on non-native contexts. It has been the escape hatch to get back to native context. For example, this is how you can relaunch native x64 cmd from the emulated x86 cmd on Windows x64:

```
C:\Windows>set PROCESSOR_ARCHITECTURE
PROCESSOR_ARCHITECTURE=x86

C:\Windows>%windir%\sysnative\cmd.exe
Microsoft Windows [Version 10.0.22631.3810]
(c) Microsoft Corporation. All rights reserved.

C:\Windows>set PROCESSOR_ARCHITECTURE
PROCESSOR_ARCHITECTURE=AMD64
```
Is it really the case that these same steps do not work in emulated x64 cmd on Windows arm64?

Reply:
Assuming there existed a build of `cmd.exe` or `powershell.exe` that consistently launched an arm64 process, then it would solve my issue. The only copy of `cmd.exe` I could find other than the System32 one is the one at `SysWOW64` which launches as an x86 process.

Reply:
You are right. I have done some reading on this topic. The system .exes are multi-arch on Windows Arm, and so there is no way to control the system process architecture by selecting different .exes. Instead, the process launching the .exe controls the process architecture. 

For example, `start` command allows you to set the desired process architecture.`start /B /WAIT /MACHINE arm64 %windir%\System32\cmd.exe` should always launch native cmd prompt on Windows arm64.

Reply:
Thanks, the `/machine` parameter seems to work for me. By installing the following script called `powershell` in my PATH:

```bash
#!/usr/bin/env bash
start //b //wait //machine arm64 powershell.exe "$@"
```

I'm able to force native powershell process from the mingw environment. This solves the problem using `build.sh` as the entrypoint, but not for the `build.cmd` entrypoint.